# Lab 4 - Group ID 4


1.   Ahmed Hannan - 23030026
2.   Hammad Javed - 23030010


## This notebook is part of the Apache Spark training delivered by CERN IT


Run this notebook from Jupyter with Python kernel
- When running on CERN SWAN, do not attach the notebook to a Spark cluster, but rather run it locally on the SWAN container (which is the default)
- If running this outside CERN SWAN, please make sure to have PySpark installed: `pip install pyspark`


In order to run this notebook as slides:
 - on SWAN click on the button "Enter/Exit RISE slideshow" in the ribbon
 - on other environments please make sure to have the RISE extension installed `pip install RISE`

### SPARK DataFrame Hands-On Lab
Contact: Luca.Canali@cern.ch

### Objective: Perform Basic DataFrame Operations
1. Creating DataFrames
2. Select columns
3. Add, rename and drop columns
4. Filtering rows
5. Aggregations

## Getting started: create the SparkSession

In [ ]:
!pip install pyspark

In [ ]:
# !pip install pyspark

from pyspark.sql import SparkSession

spark = (SparkSession.builder
          .master("local[*]") \
          .appName("DataFrame HandsOn 1") \
          .config("spark.ui.showConsoleProgress","false") \
          .getOrCreate()
        )

spark

The master `local[*]` means that the executors are in the same node that is running the driver. The `*` tells Spark to start as many executors as there are logical cores available

### Hands-On 1 - Construct a DataFrame from csv file
This demostrates how to read a csv file and construct a DataFrame.  
We will use the online retail dataset from Kaggle, credits: https://www.kaggle.com/datasets/vijayuv/onlineretail


#### First, let's inspect the csv content

In [ ]:
#modify below code to use the downloaded dataset

!gzip -cd ../data/online-retail-dataset.csv.gz 2>&1| head -n3

gzip: ../data/online-retail-dataset.csv.gz: No such file or directory


In [ ]:
!pip install kaggle

In [ ]:
!kaggle datasets download -d vijayuv/onlineretail
!unzip onlineretail.zip

Dataset URL: https://www.kaggle.com/datasets/vijayuv/onlineretail
License(s): CC0-1.0
 69% 5.00M/7.20M [00:00<00:00, 34.6MB/s]
100% 7.20M/7.20M [00:00<00:00, 47.9MB/s]
Archive:  onlineretail.zip
  inflating: OnlineRetail.csv        


In [ ]:
!mv /content/OnlineRetail.csv /content/online_retail_dataset.csv

In [ ]:
online_retail_schema="InvoiceNo int, StockCode string, Description string, Quantity int,\
InvoiceDate timestamp,UnitPrice float,CustomerId int, Country string"

In [ ]:
df = spark.read.csv("online_retail_dataset.csv", header=True, schema=online_retail_schema)
df.show(5)

+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
|InvoiceNo|StockCode|         Description|Quantity|InvoiceDate|UnitPrice|CustomerId|       Country|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|       NULL|     2.55|     17850|United Kingdom|
|   536365|    71053| WHITE METAL LANTERN|       6|       NULL|     3.39|     17850|United Kingdom|
|   536365|   84406B|CREAM CUPID HEART...|       8|       NULL|     2.75|     17850|United Kingdom|
|   536365|   84029G|KNITTED UNION FLA...|       6|       NULL|     3.39|     17850|United Kingdom|
|   536365|   84029E|RED WOOLLY HOTTIE...|       6|       NULL|     3.39|     17850|United Kingdom|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
only showing top 5 rows



#### Inspect the data

In [ ]:
df.show(2, False)

+---------+---------+----------------------------------+--------+-----------+---------+----------+--------------+
|InvoiceNo|StockCode|Description                       |Quantity|InvoiceDate|UnitPrice|CustomerId|Country       |
+---------+---------+----------------------------------+--------+-----------+---------+----------+--------------+
|536365   |85123A   |WHITE HANGING HEART T-LIGHT HOLDER|6       |NULL       |2.55     |17850     |United Kingdom|
|536365   |71053    |WHITE METAL LANTERN               |6       |NULL       |3.39     |17850     |United Kingdom|
+---------+---------+----------------------------------+--------+-----------+---------+----------+--------------+
only showing top 2 rows



#### Show columns

In [ ]:
df.printSchema()

root
 |-- InvoiceNo: integer (nullable = true)
 |-- StockCode: string (nullable = true)
 |-- Description: string (nullable = true)
 |-- Quantity: integer (nullable = true)
 |-- InvoiceDate: timestamp (nullable = true)
 |-- UnitPrice: float (nullable = true)
 |-- CustomerId: integer (nullable = true)
 |-- Country: string (nullable = true)



### Hands-On 2 - Spark Transformations - select, add, rename and drop columns

Select dataframe columns

In [ ]:
# select single column

df.select("Country").show(2)

+--------------+
|       Country|
+--------------+
|United Kingdom|
|United Kingdom|
+--------------+
only showing top 2 rows



Select multiple columns


In [ ]:
df.select("StockCode","Description","UnitPrice").show(n=2, truncate=False)

+---------+----------------------------------+---------+
|StockCode|Description                       |UnitPrice|
+---------+----------------------------------+---------+
|85123A   |WHITE HANGING HEART T-LIGHT HOLDER|2.55     |
|71053    |WHITE METAL LANTERN               |3.39     |
+---------+----------------------------------+---------+
only showing top 2 rows



In [ ]:
df.columns

['InvoiceNo',
 'StockCode',
 'Description',
 'Quantity',
 'InvoiceDate',
 'UnitPrice',
 'CustomerId',
 'Country']

In [ ]:
# select first 5 columns
df.select(df.columns[0:5]).show(2)

+---------+---------+--------------------+--------+-----------+
|InvoiceNo|StockCode|         Description|Quantity|InvoiceDate|
+---------+---------+--------------------+--------+-----------+
|   536365|   85123A|WHITE HANGING HEA...|       6|       NULL|
|   536365|    71053| WHITE METAL LANTERN|       6|       NULL|
+---------+---------+--------------------+--------+-----------+
only showing top 2 rows



In [ ]:
# selects all the original columns and adds a new column that specifies high value item
(df.selectExpr(
   "*", # all original columns
   "(UnitPrice > 100) as HighValueItem")
   .show(2)
)

+---------+---------+--------------------+--------+-----------+---------+----------+--------------+-------------+
|InvoiceNo|StockCode|         Description|Quantity|InvoiceDate|UnitPrice|CustomerId|       Country|HighValueItem|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+-------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|       NULL|     2.55|     17850|United Kingdom|        false|
|   536365|    71053| WHITE METAL LANTERN|       6|       NULL|     3.39|     17850|United Kingdom|        false|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+-------------+
only showing top 2 rows



In [ ]:
# selects all the original columns and adds a new column that specifies high value item
(df.selectExpr(
  "sum(Quantity) as TotalQuantity",
  "cast(sum(UnitPrice) as int) as InventoryValue")
  .show()
)

+-------------+--------------+
|TotalQuantity|InventoryValue|
+-------------+--------------+
|      5176450|       2498803|
+-------------+--------------+



#### Adding, renaming and dropping columns

In [ ]:
# add a new column called InvoiceValue
from pyspark.sql.functions import expr
df_1 = (df
        .withColumn("InvoiceValue", expr("UnitPrice * Quantity"))
        .select("InvoiceNo","Description","UnitPrice","Quantity","InvoiceValue")
       )
df_1.show(2, False)

# rename InvoiceValue to LineTotal
df_2 = df_1.withColumnRenamed("InvoiceValue","LineTotal")
df_2.show(2, False)

# drop a column
df_2.drop("LineTotal").show(2, False)

+---------+----------------------------------+---------+--------+------------+
|InvoiceNo|Description                       |UnitPrice|Quantity|InvoiceValue|
+---------+----------------------------------+---------+--------+------------+
|536365   |WHITE HANGING HEART T-LIGHT HOLDER|2.55     |6       |15.299999   |
|536365   |WHITE METAL LANTERN               |3.39     |6       |20.34       |
+---------+----------------------------------+---------+--------+------------+
only showing top 2 rows

+---------+----------------------------------+---------+--------+---------+
|InvoiceNo|Description                       |UnitPrice|Quantity|LineTotal|
+---------+----------------------------------+---------+--------+---------+
|536365   |WHITE HANGING HEART T-LIGHT HOLDER|2.55     |6       |15.299999|
|536365   |WHITE METAL LANTERN               |3.39     |6       |20.34    |
+---------+----------------------------------+---------+--------+---------+
only showing top 2 rows

+---------+---------

### Hands-On 3 - Spark Transformations - filter, sort and cast

In [ ]:
from pyspark.sql.functions import col

# select invoice lines with quantity > 50 and unitprice > 20
df.where(col("Quantity") > 20).where(col("UnitPrice") > 50).show(2)
df.filter(df.Quantity > 20).filter(df.UnitPrice > 50).show(2)
df.filter("Quantity > 20 and UnitPrice > 50").show(2)

+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
|InvoiceNo|StockCode|         Description|Quantity|InvoiceDate|UnitPrice|CustomerId|       Country|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
|   556444|    22502|PICNIC BASKET WIC...|      60|       NULL|    649.5|     15098|United Kingdom|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+

+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
|InvoiceNo|StockCode|         Description|Quantity|InvoiceDate|UnitPrice|CustomerId|       Country|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
|   556444|    22502|PICNIC BASKET WIC...|      60|       NULL|    649.5|     15098|United Kingdom|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+

In [ ]:
# select invoice lines with quantity > 100 or unitprice > 20
df.where((col("Quantity") > 100) | (col("UnitPrice") > 20)).show(2)

+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
|InvoiceNo|StockCode|         Description|Quantity|InvoiceDate|UnitPrice|CustomerId|       Country|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
|   536378|    21212|PACK OF 72 RETROS...|     120|       NULL|     0.42|     14688|United Kingdom|
|     NULL|        D|            Discount|      -1|       NULL|     27.5|     14527|United Kingdom|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
only showing top 2 rows



In [ ]:
from pyspark.sql.functions import desc, asc

# sort in the default order: ascending
df.orderBy(expr("UnitPrice")).show(2)

df.orderBy(col("Quantity").desc(), col("UnitPrice").asc()).show(10)

+---------+---------+---------------+--------+-----------+---------+----------+--------------+
|InvoiceNo|StockCode|    Description|Quantity|InvoiceDate|UnitPrice|CustomerId|       Country|
+---------+---------+---------------+--------+-----------+---------+----------+--------------+
|     NULL|        B|Adjust bad debt|       1|       NULL|-11062.06|      NULL|United Kingdom|
|     NULL|        B|Adjust bad debt|       1|       NULL|-11062.06|      NULL|United Kingdom|
+---------+---------+---------------+--------+-----------+---------+----------+--------------+
only showing top 2 rows

+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
|InvoiceNo|StockCode|         Description|Quantity|InvoiceDate|UnitPrice|CustomerId|       Country|
+---------+---------+--------------------+--------+-----------+---------+----------+--------------+
|   581483|    23843|PAPER CRAFT , LIT...|   80995|       NULL|     2.08|     16446|United Kingdom|
|   5

### Hands-On 4 - Spark Transformations - aggregations
full list of built int functions - https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql.html#functions

In [ ]:
%%time
# Count distinct customers
from pyspark.sql.functions import countDistinct
df.select(countDistinct("CustomerID")).show()

+--------------------------+
|count(DISTINCT CustomerID)|
+--------------------------+
|                      4372|
+--------------------------+

CPU times: user 14.4 ms, sys: 948 µs, total: 15.3 ms
Wall time: 2.56 s


In [ ]:
%%time
# approx. distinct stock items
from pyspark.sql.functions import approx_count_distinct
df.select(approx_count_distinct("CustomerID", 0.1)).show()

+---------------------------------+
|approx_count_distinct(CustomerID)|
+---------------------------------+
|                             4336|
+---------------------------------+

CPU times: user 9.32 ms, sys: 136 µs, total: 9.46 ms
Wall time: 1.62 s


In [ ]:
# average, maximum and minimum purchase quantity
from pyspark.sql.functions import avg, max, min
( df.select(
    avg("Quantity").alias("avg_purchases"),
    max("Quantity").alias("max_purchases"),
    min("Quantity").alias("min_purchases"))
   .show()
)

+----------------+-------------+-------------+
|   avg_purchases|max_purchases|min_purchases|
+----------------+-------------+-------------+
|9.55224954743324|        80995|       -80995|
+----------------+-------------+-------------+



### Hands-On 5 - Spark Transformations - grouping and windows

In [ ]:
# count of items on the invoice
df.groupBy("InvoiceNo", "CustomerId").count().show(5)

# grouping with expressions
df.groupBy("InvoiceNo").agg(expr("avg(Quantity)"),expr("stddev_pop(Quantity)"))\
  .show(5)

+---------+----------+-----+
|InvoiceNo|CustomerId|count|
+---------+----------+-----+
|   536573|     17025|    4|
|   537228|     17677|    1|
|   537419|     13495|   14|
|   538093|     12682|   33|
|   538648|     17937|    5|
+---------+----------+-----+
only showing top 5 rows

+---------+------------------+--------------------+
|InvoiceNo|     avg(Quantity)|stddev_pop(Quantity)|
+---------+------------------+--------------------+
|   536532| 25.36986301369863|  16.850272831671976|
|   537632|               1.0|                 0.0|
|   538708| 10.61111111111111|   7.150282736359209|
|   538877|14.258278145695364|   27.56989037543246|
|   538993| 9.333333333333334|   2.748737083745107|
+---------+------------------+--------------------+
only showing top 5 rows



### Read the csv file into DataFrame

`%%time` is an iPython magic https://ipython.readthedocs.io/en/stable/interactive/magics.html


It's possible to read files without specifying the schema. Some file formats (Parquet is one of them) include the schema, which means that Spark can start reading the file. For format without schema (csv, json...) Spark can infer the schema. Let's see what's the difference in terms of time and of results:

In [ ]:
online_retail_schema="InvoiceNo int, StockCode string, Description string, Quantity int,\
InvoiceDate timestamp,UnitPrice float,CustomerId int, Country string"

In [ ]:
%%time
df = spark.read \
.option("header", "true") \
.option("timestampFormat", "M/d/yyyy H:m").csv("online_retail_dataset.csv", schema=online_retail_schema)

CPU times: user 2.14 ms, sys: 875 µs, total: 3.01 ms
Wall time: 28.6 ms


In [ ]:
%%time
df_infer = spark.read \
.option("header", "true") \
.option("inferSchema", "true") \
.csv("online_retail_dataset.csv")

CPU times: user 11.4 ms, sys: 3.04 ms, total: 14.4 ms
Wall time: 3.1 s


## Exercises

Reminder: documentation at
https://spark.apache.org/docs/latest/api/python/index.html

If you didn't run the previous cells, run the following one:

In [ ]:
from pyspark.sql import SparkSession

spark = SparkSession.builder \
        .master("local[*]") \
        .appName("DataFrame HandsOn 1") \
        .config("spark.ui.showConsoleProgress","false") \
        .getOrCreate()

online_retail_schema="InvoiceNo int, StockCode string, Description string, Quantity int,\
InvoiceDate timestamp,UnitPrice float,CustomerId int, Country string"

df = spark.read \
        .option("header", "true") \
        .option("timestampFormat", "M/d/yyyy H:m")\
        .csv("../data/online-retail-dataset.csv.gz",
             schema=online_retail_schema)

Task: Show 5 lines of the "description" column

In [ ]:
df.select("Description").show(5, False)

+-----------------------------------+
|Description                        |
+-----------------------------------+
|WHITE HANGING HEART T-LIGHT HOLDER |
|WHITE METAL LANTERN                |
|CREAM CUPID HEARTS COAT HANGER     |
|KNITTED UNION FLAG HOT WATER BOTTLE|
|RED WOOLLY HOTTIE WHITE HEART.     |
+-----------------------------------+
only showing top 5 rows



✅ **Top 5** descriptions in the dataset.

Task: Count the number of distinct invoices in the dataframe

In [ ]:
from pyspark.sql.functions import countDistinct
df.select(countDistinct("InvoiceNo")).show()

+-------------------------+
|count(DISTINCT InvoiceNo)|
+-------------------------+
|                    22061|
+-------------------------+



✅ **22061** distinct invoices in the dataset.

Task: Find out in which month most invoices have been issued

In [ ]:
from pyspark.sql.functions import month
df.groupBy(month("InvoiceDate")).count().orderBy("count", ascending=False).show(3)

+------------------+-----+
|month(InvoiceDate)|count|
+------------------+-----+
|                11|84711|
|                12|68006|
|                10|60742|
+------------------+-----+
only showing top 3 rows



✅ **November** is the month with most iinvoices, followed by December & October

Task: Filter the lines where the Quantity is more than 30

In [ ]:
df.filter(df.Quantity > 30).show(n=10)

+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+
|InvoiceNo|StockCode|         Description|Quantity|        InvoiceDate|UnitPrice|CustomerId|       Country|
+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+
|   536367|    84879|ASSORTED COLOUR B...|      32|2010-12-01 08:34:00|     1.69|     13047|United Kingdom|
|   536370|    10002|INFLATABLE POLITI...|      48|2010-12-01 08:45:00|     0.85|     12583|        France|
|   536370|    22492|MINI PAINT SET VI...|      36|2010-12-01 08:45:00|     0.65|     12583|        France|
|   536371|    22086|PAPER CHAIN KIT 5...|      80|2010-12-01 09:00:00|     2.55|     13748|United Kingdom|
|   536374|    21258|VICTORIAN SEWING ...|      32|2010-12-01 09:09:00|    10.95|     15100|United Kingdom|
|   536376|    22114|HOT WATER BOTTLE ...|      48|2010-12-01 09:32:00|     3.45|     15291|United Kingdom|
|   536376|    21733|RED HAN

✅ Displaying top 10 results with **quantity greater than 30.**

Task: Show the four most sold items (by quantity)

In [ ]:
from pyspark.sql.functions import sum, desc
df.groupBy("StockCode").agg(sum("Quantity").alias("TotalQuantity")).orderBy(desc("TotalQuantity")).show(4)

+---------+-------------+
|StockCode|TotalQuantity|
+---------+-------------+
|    22197|        56450|
|    84077|        53847|
|   85099B|        47363|
|   85123A|        38830|
+---------+-------------+
only showing top 4 rows



In [ ]:
from pyspark.sql.functions import sum, desc
df.groupBy("Description").agg(sum("Quantity").alias("TotalQuantity")).orderBy(desc("TotalQuantity")).show(4)

+--------------------+-------------+
|         Description|TotalQuantity|
+--------------------+-------------+
|WORLD WAR 2 GLIDE...|        53847|
|JUMBO BAG RED RET...|        47363|
|ASSORTED COLOUR B...|        36381|
|      POPCORN HOLDER|        36334|
+--------------------+-------------+
only showing top 4 rows



✅ **Top 4 Selling Items** distinct invoices grouped by StockCode & Description. The difference in numbers shows that at some point descriptions of products were modified. So same stock exists with different descriptions.

Bonus question: why do these two operations return different results? Hint: look at the documentation

In [ ]:
print(df.select("InvoiceNo").distinct().count())
from pyspark.sql.functions import countDistinct
df.select(countDistinct("InvoiceNo")).show()

22062
+-------------------------+
|count(DISTINCT InvoiceNo)|
+-------------------------+
|                    22061|
+-------------------------+



✅ The primay difference is how they handle Null values. `.distinct().count()`  counts the Null vales while `countDistinct("InvoiceNo")` removes the null values from the count. That's why they differ with 1.